chat-plugins/tickets.js

```diff
@@ -0,0 +1,487 @@
+'use strict';
+
+const FS = require('../fs');
+const TICKET_CACHE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+const TICKET_BAN_DURATION = 48 * 60 * 60 * 1000; // 48 hours
+let tickets = {};
+
+try {
+	tickets = JSON.parse(FS('config/tickets.json').readSync());
+	for (let t in tickets) {
+		if (tickets[t].created + TICKET_CACHE_TIME <= Date.now() && !tickets[t].open) delete tickets[t];
+		if (tickets[t].open && process.uptime() <= 60) tickets[t].open = false; // Close open tickets if the server has been running for less than a minute
+	}
+} catch (e) {
+	if (e.code !== 'ENOENT') throw e;
+}
+
+function writeTickets() {
+	FS('config/tickets.json').writeUpdate(() => (
+		JSON.stringify(tickets)
+	));
+}
+
+class HelpTicket extends Rooms.RoomGame {
+	constructor(room, ticket) {
+		super(room);
+		this.ticket = ticket;
+		this.claimQueue = [];
+	}
+
+	onJoin(user, connection) {
+		if (!user.isStaff || !this.ticket.open) return false;
+		if (this.ticket.escalated && !user.can('declare')) return false;
+		if (!this.ticket.claimed) {
+			this.ticket.claimed = user.name;
+			tickets[this.ticket.userid] = this.ticket;
+			writeTickets();
+			this.room.add(`${user.name} claimed this ticket.`).update();
+			this.modnote(`${user.name} claimed this ticket.`);
+			this.updateTicket('claim');
+		} else {
+			this.claimQueue.push(user.name);
+		}
+	}
+
+	onLeave(user, connection) {
+		if (!user.isStaff || !this.ticket.open) return false;
+		if (toId(this.ticket.claimed) === user.userid) {
+			if (this.claimQueue.length) {
+				this.claimed = this.claimQueue.shift();
+				this.room.add(`This ticket is now claimed by ${this.claimed}.`).update();
+				this.modnote(`This ticket is now claimed by ${this.claimed}.`);
+			} else {
+				this.claimed = null;
+				this.room.add(`This ticket is no longer claimed.`).update();
+				this.modnote(`This ticket is no longer claimed.`);
+				this.updateTicket('unclaim');
+			}
+			tickets[this.ticket.userid] = this.ticket;
+			writeTickets();
+		} else {
+			let index = this.claimQueue.map(x => { return toId(x); }).indexOf(user.userid);
+			if (index > -1) this.claimQueue.splice(index, 1);
+		}
+	}
+
+	updateTicket(action, staff = 'A Staff member') {
+		const staffRoom = this.ticket.escalated ? Rooms('upperstaff') : Rooms('staff');
+		if (!staffRoom) return;
+		switch (action) {
+		case 'init':
+			staffRoom.add(`|uhtml|help-ticket-${this.ticket.userid}|<b>[Ticket] ${this.ticket.creator} opened a new ticket. (Type: ${this.ticket.type})</b> <button class="button" name="send" value="/join help-${this.ticket.userid}">Claim Ticket</button>`).update();
+			break;
+		case 'claim':
+			staffRoom.add(`|uhtmlchange|help-ticket${this.ticket.userid}|[Ticket] ${this.ticket.creator} opened a new ticket. (Type: ${this.ticket.type})</b> <button class="button" name="send" value="/join help-${this.ticket.userid}">Respond</button>`).update();
```

Comment:
is there intentionally no hyphen between `help-ticket` and `${this.ticket.userid}`?

Reply:
no good catch